lib/src/iterable_extensions.dart

```diff
@@ -549,6 +550,54 @@ extension IterableExtension<T> on Iterable<T> {
     }
     return true;
   }
+
+  /// Returns an iterable whose elements are contiguous slices of [this].
+  ///
+  /// Each slice is [length] elements long, except for the last one which may be
+  /// shorter if [this] contains too few elements. Each slice begins after the
+  /// last one ends.
+  ///
+  /// For example, `(1, 2, 3, 4, 5).slices(2)` returns `((1, 2), (3, 4), (5))`.
+  Iterable<Iterable<T>> slices(int length) sync* {
+    if (length < 1) throw RangeError.range(length, 1, null, 'length');
+
+    var iterator = this.iterator;
+    while (iterator.moveNext()) {
+      yield () sync* {
+        var i = 0;
+        do {
+          yield iterator.current;
+          i++;
+        } while (i < length && iterator.moveNext());
+      }();
+    }
+  }
+
+  /// Returns an iterable of each subsequence of [this] of the given [length].
+  ///
+  /// Each subsequence is exactly [length] elements long and shares `length - 1`
+  /// elements with the previous subsequence. Returns an empty iterable if
+  /// [length] is greater than this iterable's length.
+  ///
+  /// For example, `(1, 2, 3, 4).subsequences(2)` returns
+  /// `([1, 2], [2, 3], [3, 4])`.
+  Iterable<List<T>> subsequences(int length) sync* {
```

Comment:
> What is the use-case for this operation. It seems fairly inefficient.

It's useful with `length = 2` for things like creating connections between a list of elements (for example, I want to use it to connect the stdout and stdin streams of a sequence of process objects). It allows calling code to avoid manually handling the fencepost problem.

If you don't think it's worth adding, I'm not opposed to removing it. I definitely think `slices()` is the more widely applicable API, this is just the natural dual so it made sense to me to include both at once.

> Could you collect the elements into a single list, then emit views on that list. It will keep older elements alive (but so will the view used by the `List` implementation)

I think the worst-case problems with unexpectedly retaining the entire iterable contents in memory are likely to be a bigger problem than copying elements each time, especially because `length` is likely to usually be very small.